src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc

```diff
@@ -107,6 +109,29 @@ void SQLiteVersionedObjects::store_versioned_object(
   storage.update(object);
 }
 
+bool SQLiteVersionedObjects::store_versioned_object_if_state(
+    const DBVersionedObject& object, std::vector<ObjectState> allowed_states
+) const {
+  auto storage = conn->get_storage();
+  auto transaction = storage.transaction_guard();
+  const std::vector<int> version_in_allowed_state = storage.select(
+      count(&DBVersionedObject::id),
+      where(
+          in(&DBVersionedObject::object_state, allowed_states) and
+          is_equal(&DBVersionedObject::id, object.id)
+      )
+  );
+  if (version_in_allowed_state[0] == 0) {
+    // TODO(https://github.com/aquarist-labs/s3gw/issues/563) error handling
+    // throw exception (will be caught later in the sfs logic)
+    transaction.rollback();
+    return false;
+  }
```

Comment:
Agree that this is not the best solution. Follow up commit "rgw/sfs: conditional store: use update_all instead of select+update" changes this to update_all